pkg/apis/resource/structured/namedresources/types.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namedresources
+
+import "k8s.io/apimachinery/pkg/api/resource"
+
+// Resources is used in NodeResourceModel.
+type Resources struct {
+	Instances []Instance
+}
+
+type Instance struct {
+	// Name is unique identifier among all resource instances managed by
+	// the driver on the node. It must be a DNS subdomain.
+	Name string
+
+	// Attributes defines the attributes of this resource instance.
+	// The name of each attribute must be unique.
+	Attributes []Attribute
+}
+
+type Attribute struct {
+	// Name is unique identifier among all resource instances managed by
+	// the driver on the node. It must be a DNS subdomain.
+	Name string
+
+	AttributeValue
+}
+
+// AttributeValue must have one and only one field set.
+type AttributeValue struct {
+	QuantityValue    *resource.Quantity
+	BoolValue        *bool
+	IntValue         *int64
+	IntSliceValue    *IntSlice
+	StringValue      *string
+	StringSliceValue *StringSlice
+	// TODO: Version     *SemVersion
+}
+
+type IntSlice struct {
+	Ints []int64
+}
+
+type StringSlice struct {
+	Strings []string
+}
+
+// TODO
+//
+// A wrapper around https://pkg.go.dev/github.com/blang/semver/v4#Version which
+// is encoded as a string. During decoding, it validates that the string
+// can be parsed using tolerant parsing (currently trims spaces, removes a "v" prefix,
+// adds a 0 patch number to versions with only major and minor components specified,
+// and removes leading 0s).
+// type SemVersion struct {
+//	semver.Version
+//}
```

Comment:
Why is this a TODO? Meaning, what is blocking it? We will need some notion of semantic version comparisons in this initial implementation. Just using a string field and passing it to a CEL function that is aware of semver would be just as good though.

Reply:
The same could actually be said of `QuantityValue` as well, technically. I guess it really depends where we want to the validation.

Reply:
We wouldn't embed a type we don't control into our API surface. A string field we strictly define the format of is reasonable, even if that format is 1:1 with current semver.Version

We want to avoid accidentally expanding or breaking our surface area if we bump a semver dependency

Reply:
If we need to avoid parsing semver, then a major/minor/patch tuple of integers would be a simpler alternative. It's not quite the same, but perhaps good enough, @klueska?

Reply:
> So what do you suggest? Write our own semver library?

Probably?

Something like:
* make the serialization _really_ simple and dumb (normal string or typedef'ed string field)
* avoid normalization-on-roundtrip (the value you pass in must be valid to be persisted, and the value you get back is the exact same bytes)
* have Parse() and Validate() helper functions (where Validate() probably calls Parse() and just returns the error if there was one)

Reply:
> make the serialization really simple and dumb (normal string or typedef'ed string field)

Check, for the current implementation.

> avoid normalization-on-roundtrip (the value you pass in must be valid to be persisted, and the value you get back is the exact same bytes)

Not currently done, conversion is tolerant (e.g. accepts "v1.0.0") and normalizes to the spec string ("1.0.0"). Parsing could be made strict, but not re-encoding would make the Go type less useful. I prefer the approach that `resource.Quantity` is taking where the in-memory type is already usable.

> have Parse() and Validate() helper functions (where Validate() probably calls Parse() and just returns the error if there was one)

That's where it gets ugly: I don't think we should write our own semver library, and if we use semver/v4 we have the dependency problem.

Reply:
>> make the serialization really simple and dumb (normal string or typedef'ed string field)
>
> Check, for the current implementation.

needing to do custom json and proto marshaling and unmarshaling isn't what I meant by simple and dumb. An actual string field is what I meant :)

resource.Quantity has been a huge headache. I don't think we should repeat that approach.
* custom marshaling/unmarshaling means a format error fails decode, rather than failing validation (where all the validation errors can be returned at once)
* normalization on write / read roundtrip has broken many things in server-side apply

Reply:
Ack.

So would "plain string" + semver/v4 for validation and runtime version comparison be okay?

Reply:
I'm still interested in the answer, but for now I've taken the semver commits out of the PR again. Let's focus on getting the rest reviewed and merged first.

Reply:
Let's continue in https://github.com/kubernetes/kubernetes/pull/123664.

That PR extends the CEL library and could be merged separately from this PR.

Reply:
When I discussed this with @klueska, he pointed out that semver comparison is important for them. Therefore let's try to get this included in 1.30.

We can continue to discuss inclusion in the base CEL env in https://github.com/kubernetes/kubernetes/pull/123664, but in the meantime I have included one more commit here which is less intrusive and less risky:
- The API type is a simple string.
- Validation is using the regular expression from semver.org, not the semver/v4 implementation.
- The CEL type is only getting added to the DRA CEL environment, not the base environment, which makes it alpha quality.

cc @cici37